File: crawler/security/transport_budget.py
"""Fixed transport budget and proxy contract for TASK-022D-1.

This module is pure decision logic. It performs no I/O, does not read proxy
environment variables, and never creates sockets or DNS requests.
"""

from __future__ import annotations

from dataclasses import dataclass
from types import MappingProxyType


class BudgetError(ValueError):
    def __init__(self, reason_code: str, message: str = "") -> None:
        super().__init__(message or reason_code)
        self.reason_code = reason_code


@dataclass(frozen=True, slots=True)
class TransportBudget:
    dns_timeout_ms: int = 5000
    connect_timeout_ms: int = 5000
    tls_timeout_ms: int = 5000
    response_header_timeout_ms: int = 10000
    read_idle_timeout_ms: int = 15000
    request_body_bytes: int = 1048576
    response_headers_bytes: int = 262144
    max_redirects: int = 3
    global_active: int = 20
    per_host_active: int = 5
    per_host_idle: int = 2


@dataclass(frozen=True, slots=True)
class BudgetProfile:
    name: str
    total_timeout_ms: int
    response_body_bytes: int


BUDGET = TransportBudget()
PROFILES = MappingProxyType(
    {
        "probe": BudgetProfile("probe", 30000, 1048576),
        "search": BudgetProfile("search", 30000, 8388608),
        "detail": BudgetProfile("detail", 60000, 20971520),
    }
)
MAX_REDIRECTS = BUDGET.max_redirects


def _as_int(value: int, *, field: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise BudgetError("invalid_budget_value")
    return value


def budget_for(profile: str) -> BudgetProfile:
    if profile not in PROFILES:
        raise BudgetError("unknown_profile")
    return PROFILES[profile]


def remaining_deadline(profile: str, elapsed_ms: int) -> int:
    selected = budget_for(profile)
    elapsed = _as_int(elapsed_ms, field="elapsed_ms")
    if elapsed < 0:
        raise BudgetError("negative_budget")
    if elapsed > selected.total_timeout_ms:
        raise BudgetError("total_timeout_exceeded")
    return selected.total_timeout_ms - elapsed


def capped_stage_timeout(profile: str, stage_timeout_ms: int, elapsed_ms: int) -> int:
    stage = _as_int(stage_timeout_ms, field="stage_timeout_ms")
    if stage <= 0:
        raise BudgetError("non_positive_budget")
    remaining = remaining_deadline(profile, elapsed_ms)
    if remaining == 0:
        raise BudgetError("total_timeout_exceeded")
    return min(stage, remaining)


def check_byte_limit(value: int, limit: int, *, reason_code: str) -> None:
    count = _as_int(value, field="value")
    if count < 0:
        raise BudgetError("negative_budget")
    if count > limit:
        raise BudgetError(reason_code)


def check_request_body_size(value: int) -> None:
    check_byte_limit(value, BUDGET.request_body_bytes, reason_code="request_body_limit")


def check_response_headers_size(value: int) -> None:
    check_byte_limit(value, BUDGET.response_headers_bytes, reason_code="response_headers_limit")


def check_response_body_size(profile: str, value: int) -> None:
    selected = budget_for(profile)
    check_byte_limit(value, selected.response_body_bytes, reason_code="response_body_limit")


def validate_proxy(proxy: str | None) -> None:
    if proxy is None or proxy == "":
        return
    raise BudgetError("proxy_not_allowed")

File: go-spider/internal/security/transport_budget.go
package security

// BudgetError is a stable, sanitized budget/proxy policy failure.
type BudgetError struct {
	ReasonCode string
}

func (e *BudgetError) Error() string { return e.ReasonCode }

func budgetFail(reason string) error { return &BudgetError{ReasonCode: reason} }

const (
	DNSTimeoutMS            = 5000
	ConnectTimeoutMS        = 5000
	TLSTimeoutMS            = 5000
	ResponseHeaderTimeoutMS = 10000
	ReadIdleTimeoutMS       = 15000
	RequestBodyBytes        = 1048576
	ResponseHeadersBytes    = 262144
	MaxRedirects            = 3
	GlobalActive            = 20
	PerHostActive           = 5
	PerHostIdle             = 2
)

// TransportBudget is an immutable fixed budget model.
type TransportBudget struct {
	dnsTimeoutMS            int
	connectTimeoutMS        int
	tlsTimeoutMS            int
	responseHeaderTimeoutMS int
	readIdleTimeoutMS       int
	requestBodyBytes        int
	responseHeadersBytes    int
	maxRedirects            int
	globalActive            int
	perHostActive           int
	perHostIdle             int
}

// NewTransportBudget returns the fixed sealed budget.
func NewTransportBudget() TransportBudget {
	return TransportBudget{
		dnsTimeoutMS:            DNSTimeoutMS,
		connectTimeoutMS:        ConnectTimeoutMS,
		tlsTimeoutMS:            TLSTimeoutMS,
		responseHeaderTimeoutMS: ResponseHeaderTimeoutMS,
		readIdleTimeoutMS:       ReadIdleTimeoutMS,
		requestBodyBytes:        RequestBodyBytes,
		responseHeadersBytes:    ResponseHeadersBytes,
		maxRedirects:            MaxRedirects,
		globalActive:            GlobalActive,
		perHostActive:           PerHostActive,
		perHostIdle:             PerHostIdle,
	}
}

func (b TransportBudget) DNSTimeoutMS() int            { return b.dnsTimeoutMS }
func (b TransportBudget) ConnectTimeoutMS() int        { return b.connectTimeoutMS }
func (b TransportBudget) TLSTimeoutMS() int            { return b.tlsTimeoutMS }
func (b TransportBudget) ResponseHeaderTimeoutMS() int { return b.responseHeaderTimeoutMS }
func (b TransportBudget) ReadIdleTimeoutMS() int       { return b.readIdleTimeoutMS }
func (b TransportBudget) RequestBodyBytes() int        { return b.requestBodyBytes }
func (b TransportBudget) ResponseHeadersBytes() int    { return b.responseHeadersBytes }
func (b TransportBudget) MaxRedirects() int            { return b.maxRedirects }
func (b TransportBudget) GlobalActive() int            { return b.globalActive }
func (b TransportBudget) PerHostActive() int           { return b.perHostActive }
func (b TransportBudget) PerHostIdle() int             { return b.perHostIdle }

// BudgetProfile is an immutable request-class budget.
type BudgetProfile struct {
	name              string
	totalTimeoutMS    int
	responseBodyBytes int
}

func (p BudgetProfile) Name() string           { return p.name }
func (p BudgetProfile) TotalTimeoutMS() int    { return p.totalTimeoutMS }
func (p BudgetProfile) ResponseBodyBytes() int { return p.responseBodyBytes }

// BudgetForProfile returns the fixed profile for probe/search/detail.
func BudgetForProfile(name string) (BudgetProfile, error) {
	switch name {
	case "probe":
		return BudgetProfile{name: "probe", totalTimeoutMS: 30000, responseBodyBytes: 1048576}, nil
	case "search":
		return BudgetProfile{name: "search", totalTimeoutMS: 30000, responseBodyBytes: 8388608}, nil
	case "detail":
		return BudgetProfile{name: "detail", totalTimeoutMS: 60000, responseBodyBytes: 20971520}, nil
	default:
		return BudgetProfile{}, budgetFail("unknown_profile")
	}
}

func checkInt(value int) error {
	if value < 0 {
		return budgetFail("negative_budget")
	}
	return nil
}

// RemainingDeadline returns total - elapsed and rejects elapsed over budget.
func RemainingDeadline(profile BudgetProfile, elapsedMS int) (int, error) {
	if err := checkInt(elapsedMS); err != nil {
		return 0, err
	}
	if elapsedMS > profile.totalTimeoutMS {
		return 0, budgetFail("total_timeout_exceeded")
	}
	return profile.totalTimeoutMS - elapsedMS, nil
}

// CappedStageTimeout bounds a stage timeout by the remaining total deadline.
func CappedStageTimeout(profile BudgetProfile, stageTimeoutMS, elapsedMS int) (int, error) {
	if stageTimeoutMS <= 0 {
		return 0, budgetFail("non_positive_budget")
	}
	remaining, err := RemainingDeadline(profile, elapsedMS)
	if err != nil {
		return 0, err
	}
	if remaining <= 0 {
		return 0, budgetFail("total_timeout_exceeded")
	}
	if stageTimeoutMS < remaining {
		return stageTimeoutMS, nil
	}
	return remaining, nil
}

// CheckByteLimit accepts values up to limit inclusively.
func CheckByteLimit(value, limit int, reason string) error {
	if err := checkInt(value); err != nil {
		return err
	}
	if value > limit {
		return budgetFail(reason)
	}
	return nil
}

// CheckRequestBodyLimit validates the fixed request body budget.
func CheckRequestBodyLimit(value int) error {
	return CheckByteLimit(value, RequestBodyBytes, "request_body_limit")
}

// CheckResponseHeadersLimit validates the fixed response header budget.
func CheckResponseHeadersLimit(value int) error {
	return CheckByteLimit(value, ResponseHeadersBytes, "response_headers_limit")
}

// CheckResponseBodyLimit validates a profile response body budget.
func CheckResponseBodyLimit(profile BudgetProfile, value int) error {
	return CheckByteLimit(value, profile.responseBodyBytes, "response_body_limit")
}

// ValidateProxy rejects every non-empty proxy configuration.
func ValidateProxy(proxy string) error {
	if proxy == "" {
		return nil
	}
	return budgetFail("proxy_not_allowed")
}
